File: src/main/cpp/my_monad.cpp
#include <functional>
#include <iostream>

/* ---- ---- ---- define MyList ---- ---- ---- */
template <class T> class List {
  public:
  T value;
  List<T> *next;

  List(T v, List<T> *n) : value(v), next(n) { }
  ~List() {
    if (next != nullptr) {
      delete next;
      next = nullptr;
    }
  }

  template<class U> U foldRight (U seed, std::function<U(T, U)> f) {
    if (next == nullptr)
      return seed;
    else
      return f(value, next->foldRight(seed, f));
  }

  std::ostream &writeBuff(std::ostream &out) {
    out << value << ". ";
    if (next == nullptr)
      return out;
    else
      return next->writeBuff(out);
  }
};

template<class T> List<T> *cons(T item, List<T> *next) {
  return new List<T>(item, next);
};

/* ---- ---- ---- define Monadaa and use ---- ---- ---- */

template<template<class> class M> class Monadaa {
  template<class A, class B> static M<B> map (M<A> ma, std::function<B(A)> f);
  template<class A, class B> static M<B> bind (M<A> ma, std::function<M<B>(A)> f);
};

template<
  template<class> class M,
  class A,
  class B,
  class X
>
M<X> flat (M<A> ma, M<B> mb, std::function<X(A, B)> f) {
  return M.bind(ma, [](A a) {
    return M.map(mb, [&a](B b) {
      return f(a, b);
    });
  });
}

/* ---- ---- ---- List is Monadaa ---- ---- ---- */
template <> class Monadaa<List> {
  template<class A, class B> static List<B> map(List<A>, std::function<B(A)> f) {
    return nullptr;
  }

  template<class A, class B> static List<B> bind(List<A>, std::function<List<B>(A)> f) {
    return nullptr;
  }
};

/* ---- ---- ---- Entry point ---- ---- ---- */

int main () {
  auto *one = cons(1, cons(2, cons<int>(3, nullptr)));
  one->writeBuff(std::cout) << std::endl;

  delete one;
  return 0;
}

File: src/main/ruby/my_monad.rb

# ---- ---- ---- define List ---- ---- ----
List = Struct.new(:item, :next) do
  def fold_right(seed, &block)
    seed = self.next.fold_right seed, &block if self.next
    block.call(self.item, seed)
  end

  def inspect
    self.item.inspect + ". " + (self.next.nil? ? "" : self.next.inspect)
  end
end

# ---- ---- ---- define Monadaa ---- ---- ----

module Monadaa
  REQUIRED_METHOD = [:map, :bind]
  @regist = {}

  module_function
  def [](klass)
    @regist[klass] or raise "#{klass} is not a Monadaa instance."
  end

  def register(klass, &block)
    obj = Object.new
    obj.instance_eval &block
    REQUIRED_METHOD.each do |m|
      raise "Must implement #{m} for #{klass}" unless obj.respond_to? m
    end
    @regist[klass] = obj
  end
end

def flat(xs, ys, &block)
  m = Monadaa[xs.class]
  m.bind xs do |x|
    m.map ys do |y|
      yield x, y
    end
  end
end

# ---- ---- ---- List is Monadaa ---- ---- ----
Monadaa.register List do
  def map(xs, &block)
    xs.fold_right nil do |item, dst|
      List.new (yield item), dst
    end
  end

  def bind(xs, &block)
    xs.fold_right nil do |x, y|
      (yield x).fold_right y, &List.method(:new)
    end
  end
end

# ---- ---- ---- Execute block ---- ---- ----
begin
  one = List.new( 1, List.new( 2, List.new( 3, nil)))
  ten = List.new(10, List.new(20, List.new(30, nil)))
  p flat(one, ten, &:+)
end

